modules/chemical_reactions/src/actions/ChemicalCompositionAction.C

```diff
@@ -0,0 +1,208 @@
+  params.addParam<MooseEnum>("munit", mUnit, "Mass Unit");
+
+  params.addParam<std::vector<std::string>>("output_phases", "List of phases to be output");
+  params.addParam<std::vector<std::string>>(
+      "output_species", "List species for which concentration in the phases is needed");
+  params.addParam<std::vector<std::string>>(
+      "element_potentials",
+      "List of chemical elements for which chemical potentials are requested");
+  return params;
+}
+
+ChemicalCompositionAction::ChemicalCompositionAction(const InputParameters & params)
+  : Action(params),
+    _elements(getParam<std::vector<std::string>>("elements")),
+    _tunit(getParam<MooseEnum>("tunit")),
+    _punit(getParam<MooseEnum>("punit")),
+    _munit(getParam<MooseEnum>("munit")),
+    _phases(getParam<std::vector<std::string>>("output_phases")),
+    _species(getParam<std::vector<std::string>>("output_species")),
+    _element_potentials(getParam<std::vector<std::string>>("element_potentials"))
+{
+  std::replace(_munit.begin(), _munit.end(), '_', ' ');
+#ifndef THERMOCHIMICA_ENABLED
+  mooseError("Thermochimica disabled");
+#endif
+}
+
+void
+ChemicalCompositionAction::act()
+{
+#ifdef THERMOCHIMICA_ENABLED
+  //
+  // Add AuxVariables
+  //
+  if (_current_task == "add_aux_variable")
+  {
+    const std::string aux_var_type = "MooseVariable";
+    auto params = _factory.getValidParams(aux_var_type);
+    const bool second = _problem->mesh().hasSecondOrderElements();
+    params.set<MooseEnum>("order") = second ? "SECOND" : "FIRST";
+    params.set<MooseEnum>("family") = "LAGRANGE";
+
+    for (const auto i : index_range(_elements))
+      _problem->addAuxVariable(aux_var_type, _elements[i], params);
+
+    for (const auto i : index_range(_phases))
+      _problem->addAuxVariable(aux_var_type, _phases[i], params);
+
+    for (const auto i : index_range(_species))
+      _problem->addAuxVariable(aux_var_type, _species[i], params);
+
+    for (const auto i : index_range(_element_potentials))
+      _problem->addAuxVariable(aux_var_type, _element_potentials[i], params);
+  }
+
+  //
+  // Set up initial conditions from a file
+  //
+  if (_current_task == "add_ic" && isParamValid("initial_values"))
+  {
+    readCSV();
+    for (auto it : _initial_conditions)
+    {
+      const std::string class_name = "ConstantIC";
+      auto params = _factory.getValidParams(class_name);
+      params.set<VariableName>("variable") = it.first;
+      params.set<Real>("value") = it.second;
+      _problem->addInitialCondition(class_name, it.first + "_ic", params);
+    }
+  }
+
+  //
+  // Initiate Chemistry Model
+  //
+  if (_current_task == "add_user_object")
+  {
+    // initialize database in Thermochimica
+    if (isParamValid("thermofile"))
+    {
+      const auto thermo_file = getParam<FileName>("thermofile");
+      if (thermo_file.length() > 120)
+        paramError("thermofile",
+                   "Path exceeds thermiochimica's maximal permisible length of 120 with ",
```

Comment:
```suggestion
                   "Path exceeds thermochimica's maximal permissible length of 120 with ",
```

Reply:
ah, this was fixed in https://github.com/ORNL-CEES/thermochimica/pull/109 ...

Reply:
remove then?

Reply:
that PR was merged

Reply:
you should be able to remove this line

Reply:
changed it to 1024 (which is still an internal limit)